File: src/backend/src/main/java/com/ted/signal/application/service/SignalDetectionService.java
package com.ted.signal.application.service;

import com.ted.signal.application.port.in.DetectSignalsUseCase;
import com.ted.signal.application.port.out.*;
import com.ted.signal.domain.enums.SignalGrade;
import com.ted.signal.domain.enums.SignalType;
import com.ted.signal.domain.model.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class SignalDetectionService implements DetectSignalsUseCase {

    private static final BigDecimal RAPID_DECLINE_THRESHOLD = new BigDecimal("-10.0");
    private static final int TREND_MA_SHORT = 5;
    private static final int TREND_MA_LONG = 20;
    private static final int TREND_HISTORY_DAYS = TREND_MA_LONG + 10;
    private static final int VOLUME_HISTORY_DAYS = 30;

    private final StockRepository stockRepository;
    private final StockPriceRepository stockPriceRepository;
    private final LendingBalanceRepository lendingBalanceRepository;
    private final ShortSellingRepository shortSellingRepository;
    private final SignalRepository signalRepository;

    @Override
    @Transactional
    public DetectionResult detectAll(LocalDate date) {
        long start = System.currentTimeMillis();
        log.info("=== 시그널 탐지 시작: {} ===", date);

        // 1. 전체 데이터 벌크 로드 (N+1 제거: 종목당 7쿼리 → 전체 6쿼리)
        List<Stock> stocks = stockRepository.findByIsActiveTrueAndDeletedAtIsNull();
        Set<Long> activeStockIds = stocks.stream().map(Stock::getId).collect(Collectors.toSet());

        Map<Long, LendingBalance> balanceByStock = lendingBalanceRepository.findAllByTradingDate(date)
                .stream()
                .collect(Collectors.toMap(lb -> lb.getStock().getId(), lb -> lb, (a, b) -> a));

        Map<Long, StockPrice> priceByStock = stockPriceRepository.findAllByTradingDate(date)
                .stream()
                .collect(Collectors.toMap(sp -> sp.getStock().getId(), sp -> sp, (a, b) -> a));

        Map<Long, ShortSelling> shortByStock = shortSellingRepository.findAllByTradingDate(date)
                .stream()
                .collect(Collectors.toMap(ss -> ss.getStock().getId(), ss -> ss, (a, b) -> a));

        // 추세전환용 대차잔고 히스토리 (활성 종목만, 이후 메모리에서 정렬)
        LocalDate trendFrom = date.minusDays(TREND_HISTORY_DAYS);
        Map<Long, List<LendingBalance>> trendHistory = activeStockIds.isEmpty()
                ? Map.of()
                : lendingBalanceRepository
                        .findAllByStockIdsAndTradingDateBetween(activeStockIds, trendFrom, date)
                        .stream()
                        .collect(Collectors.groupingBy(lb -> lb.getStock().getId()));

        // 숏스퀴즈용 거래량 히스토리 (활성 종목, 당일 제외)
        LocalDate volumeFrom = date.minusDays(VOLUME_HISTORY_DAYS);
        Map<Long, List<StockPrice>> volumeHistory = activeStockIds.isEmpty()
                ? Map.of()
                : stockPriceRepository
                        .findAllByStockIdsAndTradingDateBetween(activeStockIds, volumeFrom, date.minusDays(1))
                        .stream()
                        .collect(Collectors.groupingBy(sp -> sp.getStock().getId()));

        // 기존 시그널 존재 여부 Set (JOIN FETCH로 N+1 방지)
        Set<String> existingKeys = signalRepository.findBySignalDateWithStockOrderByScoreDesc(date)
                .stream()
                .map(s -> signalKey(s.getStock().getId(), s.getSignalType()))
                .collect(Collectors.toCollection(HashSet::new));

        // 2. 메모리 기반 탐지 루프 (DB 히트 없음)
        List<Signal> toSave = new ArrayList<>();

        for (Stock stock : stocks) {
            Long stockId = stock.getId();

            detectRapidDecline(stock, date, balanceByStock.get(stockId), existingKeys)
                    .ifPresent(toSave::add);

            detectTrendReversal(stock, date, trendHistory.getOrDefault(stockId, List.of()), existingKeys)
                    .ifPresent(toSave::add);

            detectShortSqueeze(
                    stock, date,
                    balanceByStock.get(stockId),
                    priceByStock.get(stockId),
                    shortByStock.get(stockId),
                    volumeHistory.getOrDefault(stockId, List.of()),
                    existingKeys
            ).ifPresent(toSave::add);
        }

        // 3. 일괄 저장
        if (!toSave.isEmpty()) {
            signalRepository.saveAll(toSave);
        }

        long rapidCount = toSave.stream().filter(s -> s.getSignalType() == SignalType.RAPID_DECLINE).count();
        long trendCount = toSave.stream().filter(s -> s.getSignalType() == SignalType.TREND_REVERSAL).count();
        long squeezeCount = toSave.stream().filter(s -> s.getSignalType() == SignalType.SHORT_SQUEEZE).count();

        long elapsed = System.currentTimeMillis() - start;
        log.info("=== 시그널 탐지 완료: 급감 {}, 추세전환 {}, 숏스퀴즈 {} ({}ms) ===",
                rapidCount, trendCount, squeezeCount, elapsed);

        return new DetectionResult((int) rapidCount, (int) trendCount, (int) squeezeCount, elapsed);
    }

    /**
     * 시그널 1: 대차잔고 급감
     */
    private Optional<Signal> detectRapidDecline(Stock stock, LocalDate date,
                                                LendingBalance lb, Set<String> existingKeys) {
        if (lb == null || lb.getChangeRate() == null
                || lb.getChangeRate().compareTo(RAPID_DECLINE_THRESHOLD) > 0) {
            return Optional.empty();
        }

        String key = signalKey(stock.getId(), SignalType.RAPID_DECLINE);
        if (!existingKeys.add(key)) return Optional.empty();

        int score = calculateRapidDeclineScore(lb);
        var detail = Map.<String, Object>of(
                "balanceChangeRate", lb.getChangeRate(),
                "changeQuantity", lb.getChangeQuantity(),
                "consecutiveDecreaseDays", lb.getConsecutiveDecreaseDays()
        );

        return Optional.of(Signal.builder()
                .stock(stock).signalDate(date)
                .signalType(SignalType.RAPID_DECLINE)
                .score(score).grade(SignalGrade.fromScore(score))
                .detail(detail)
                .build());
    }

    /**
     * 시그널 2: 추세전환 (골든크로스)
     */
    private Optional<Signal> detectTrendReversal(Stock stock, LocalDate date,
                                                 List<LendingBalance> history, Set<String> existingKeys) {
        if (history.size() < TREND_MA_LONG + 1) return Optional.empty();

        // tradingDate 오름차순 보장 (groupingBy는 순서 보장 안 됨)
        var sorted = history.stream()
                .sorted(Comparator.comparing(LendingBalance::getTradingDate))
                .toList();

        var quantities = sorted.stream().map(LendingBalance::getBalanceQuantity).toList();
        int size = quantities.size();

        double maShortToday = movingAverage(quantities, size - 1, TREND_MA_SHORT);
        double maLongToday = movingAverage(quantities, size - 1, TREND_MA_LONG);
        double maShortYesterday = movingAverage(quantities, size - 2, TREND_MA_SHORT);
        double maLongYesterday = movingAverage(quantities, size - 2, TREND_MA_LONG);

        boolean goldenCross = maShortYesterday >= maLongYesterday && maShortToday < maLongToday;
        if (!goldenCross) return Optional.empty();

        String key = signalKey(stock.getId(), SignalType.TREND_REVERSAL);
        if (!existingKeys.add(key)) return Optional.empty();

        int score = calculateTrendReversalScore(maShortToday, maLongToday, maShortYesterday, maLongYesterday);
        var detail = Map.<String, Object>of(
                "maShort", round(maShortToday),
                "maLong", round(maLongToday),
                "crossType", "GOLDEN_CROSS"
        );

        return Optional.of(Signal.builder()
                .stock(stock).signalDate(date)
                .signalType(SignalType.TREND_REVERSAL)
                .score(score).grade(SignalGrade.fromScore(score))
                .detail(detail)
                .build());
    }

    /**
     * 시그널 3: 숏스퀴즈 종합 스코어
     */
    private Optional<Signal> detectShortSqueeze(Stock stock, LocalDate date,
                                                LendingBalance lb, StockPrice sp, ShortSelling ss,
                                                List<StockPrice> priceHistory,
                                                Set<String> existingKeys) {
        if (lb == null || sp == null) return Optional.empty();

        int balanceScore = scoreBalanceChange(lb.getChangeRate());
        BigDecimal volumeRatio = calcVolumeRatio(priceHistory, sp);
        int volumeScore = scoreVolumeChange(volumeRatio);
        int priceScore = scorePriceChange(sp.getChangeRate());
        int shortRatioScore = scoreShortRatio(ss);

        int totalScore = balanceScore + volumeScore + priceScore + shortRatioScore;
        if (totalScore < 40) return Optional.empty();

        String key = signalKey(stock.getId(), SignalType.SHORT_SQUEEZE);
        if (!existingKeys.add(key)) return Optional.empty();

        var detail = Map.<String, Object>of(
                "balanceChangeRate", lb.getChangeRate() != null ? lb.getChangeRate() : BigDecimal.ZERO,
                "balanceScore", balanceScore,
                "volumeChangeRate", volumeRatio,
                "volumeScore", volumeScore,
                "priceChangeRate", sp.getChangeRate() != null ? sp.getChangeRate() : BigDecimal.ZERO,
                "priceScore", priceScore,
                "shortRatioScore", shortRatioScore,
                "consecutiveDecreaseDays", lb.getConsecutiveDecreaseDays()
        );

        return Optional.of(Signal.builder()
                .stock(stock).signalDate(date)
                .signalType(SignalType.SHORT_SQUEEZE)
                .score(totalScore).grade(SignalGrade.fromScore(totalScore))
                .detail(detail)
                .build());
    }

    // ========== Scoring Functions ==========

    private int calculateRapidDeclineScore(LendingBalance lb) {
        double absChange = Math.abs(lb.getChangeRate().doubleValue());
        int baseScore = (int) Math.min(60, absChange * 3);
        int consecutiveBonus = Math.min(20, lb.getConsecutiveDecreaseDays() * 5);
        return Math.min(100, baseScore + consecutiveBonus + 20);
    }

    private int calculateTrendReversalScore(double maShortToday, double maLongToday,
                                            double maShortYesterday, double maLongYesterday) {
        double divergence = Math.abs(maShortToday - maLongToday) / maLongToday * 100;
        int divergenceScore = (int) Math.min(40, divergence * 10);
        double speed = Math.abs((maShortToday - maShortYesterday) / maShortYesterday * 100);
        int speedScore = (int) Math.min(30, speed * 15);
        return Math.min(100, divergenceScore + speedScore + 30);
    }

    private int scoreBalanceChange(BigDecimal changeRate) {
        if (changeRate == null) return 0;
        double absChange = Math.abs(changeRate.doubleValue());
        return (int) Math.min(30, absChange * 1.5);
    }

    /** 오늘 거래량 / 20일 평균 거래량 비율 */
    private BigDecimal calcVolumeRatio(List<StockPrice> priceHistory, StockPrice today) {
        if (priceHistory.isEmpty() || today.getVolume() == 0) return BigDecimal.ZERO;
        double avgVolume = priceHistory.stream().mapToLong(StockPrice::getVolume).average().orElse(0);
        if (avgVolume == 0) return BigDecimal.ZERO;
        return BigDecimal.valueOf(today.getVolume() / avgVolume)
                .setScale(4, RoundingMode.HALF_UP);
    }

    /** 거래량 비율 → 스코어 (max 25) */
    private int scoreVolumeChange(BigDecimal ratio) {
        if (ratio == null || ratio.signum() == 0) return 0;
        double r = ratio.doubleValue();
        return (int) Math.max(0, Math.min(25, (r - 1) * 12.5));
    }

    private int scorePriceChange(BigDecimal changeRate) {
        if (changeRate == null) return 0;
        double change = changeRate.doubleValue();
        if (change <= 0) return 0;
        return (int) Math.min(25, change * 5);
    }

    private int scoreShortRatio(ShortSelling ss) {
        if (ss == null || ss.getShortRatio() == null) return 0;
        double ratio = ss.getShortRatio().doubleValue();
        return (int) Math.min(20, ratio * 2);
    }

    // ========== Helpers ==========

    private static String signalKey(Long stockId, SignalType type) {
        return stockId + ":" + type.name();
    }

    private double movingAverage(List<Long> values, int endIndex, int period) {
        int startIdx = Math.max(0, endIndex - period + 1);
        return values.subList(startIdx, endIndex + 1).stream()
                .mapToLong(Long::longValue).average().orElse(0);
    }

    private double round(double value) {
        return BigDecimal.valueOf(value).setScale(2, RoundingMode.HALF_UP).doubleValue();
    }
}

File: src/backend_py/app/application/service/signal_detection_service.py
"""3종 시그널 탐지 — Java 로직 1:1 포팅, pandas 벡터화.

- RAPID_DECLINE: 대차잔고 change_rate <= -12% → base(abs*2.5) + consec(*5) + 10, cap 100
- TREND_REVERSAL: 대차잔고 5MA vs 20MA 골든→데드 크로스 검출(어제 >= 오늘 <), score >= 50
- SHORT_SQUEEZE: balance+volume+price+short_ratio 부분점수 합이 60 이상

2026-04-20 임계값 튜닝: 3년 백필 결과(70,609건, hit_rate ~45%)에서 저등급 비중이 과도.
SHORT_SQUEEZE C-grade 81%, TREND_REVERSAL D-grade 22%, RAPID_DECLINE A-grade 62% —
각 타입별 기준치를 재정비해 상위 신호만 남도록 조정했다.
"""
from __future__ import annotations

import logging
import time
from datetime import date, timedelta
from decimal import Decimal
from typing import Any

import pandas as pd
from sqlalchemy.ext.asyncio import AsyncSession

from app.adapter.out.persistence.models import Signal, SignalGrade, SignalType
from app.adapter.out.persistence.repositories import (
    LendingBalanceRepository,
    ShortSellingRepository,
    SignalRepository,
    StockPriceRepository,
    StockRepository,
)
from app.application.dto.results import DetectionResult

logger = logging.getLogger(__name__)

RAPID_DECLINE_THRESHOLD = Decimal("-12.0")
TREND_MA_SHORT = 5
TREND_MA_LONG = 20
TREND_HISTORY_DAYS = TREND_MA_LONG + 10
TREND_REVERSAL_MIN_SCORE = 50
VOLUME_HISTORY_DAYS = 30
SHORT_SQUEEZE_MIN_SCORE = 60


def _to_decimal(value: Any) -> Decimal | None:
    if value is None:
        return None
    try:
        return Decimal(str(value))
    except Exception:
        return None


def _grade(score: int) -> str:
    return SignalGrade.from_score(score).value


class SignalDetectionService:
    def __init__(self, session: AsyncSession) -> None:
        self._session = session

    async def detect_all(self, trading_date: date) -> DetectionResult:
        start = time.monotonic()
        logger.info("시그널 탐지 시작 date=%s", trading_date)

        stock_repo = StockRepository(self._session)
        price_repo = StockPriceRepository(self._session)
        short_repo = ShortSellingRepository(self._session)
        lending_repo = LendingBalanceRepository(self._session)
        signal_repo = SignalRepository(self._session)

        stocks = list(await stock_repo.list_active())
        if not stocks:
            return DetectionResult(rapid_decline=0, trend_reversal=0, short_squeeze=0, elapsed_ms=0)

        stock_ids = [s.id for s in stocks]

        prices_today = await price_repo.list_by_trading_date(trading_date)
        shorts_today = await short_repo.list_by_trading_date(trading_date)
        lendings_today = await lending_repo.list_by_trading_date(trading_date)

        price_by_stock = {p.stock_id: p for p in prices_today}
        short_by_stock = {s.stock_id: s for s in shorts_today}
        lending_by_stock = {lb.stock_id: lb for lb in lendings_today}

        # 추세전환용 히스토리(lending_balance 수량)
        trend_from = trading_date - timedelta(days=TREND_HISTORY_DAYS)
        lending_hist = await lending_repo.list_by_stocks_between(
            stock_ids, trend_from, trading_date
        )
        trend_df = pd.DataFrame(
            [
                {"stock_id": lb.stock_id, "date": lb.trading_date, "qty": int(lb.balance_quantity)}
                for lb in lending_hist
            ]
        )

        # 숏스퀴즈용 볼륨 히스토리 (당일 제외)
        vol_from = trading_date - timedelta(days=VOLUME_HISTORY_DAYS)
        vol_hist = await price_repo.list_by_stocks_between(
            stock_ids, vol_from, trading_date - timedelta(days=1)
        )
        vol_df = pd.DataFrame(
            [{"stock_id": sp.stock_id, "volume": int(sp.volume)} for sp in vol_hist]
        )
        avg_volume_by_stock: dict[int, float] = {}
        if not vol_df.empty:
            # groupby().mean().to_dict() 는 Hashable/Any 키로 좁혀지지 않아 cast.
            raw_mean: dict[Any, Any] = vol_df.groupby("stock_id")["volume"].mean().to_dict()
            for k, v in raw_mean.items():
                avg_volume_by_stock[int(k)] = float(v)

        # 기존 시그널 중복 방지
        existing = await signal_repo.list_by_date(trading_date)
        existing_keys = {(s.stock_id, s.signal_type) for s in existing}

        to_save: list[Signal] = []

        for stock in stocks:
            sid = stock.id
            lb = lending_by_stock.get(sid)
            sp = price_by_stock.get(sid)
            ss = short_by_stock.get(sid)
            hist = trend_df[trend_df["stock_id"] == sid] if not trend_df.empty else None

            sig = self._rapid_decline(stock.id, trading_date, lb, existing_keys)
            if sig:
                to_save.append(sig)

            sig = self._trend_reversal(stock.id, trading_date, hist, existing_keys)
            if sig:
                to_save.append(sig)

            sig = self._short_squeeze(
                stock.id, trading_date, lb, sp, ss,
                avg_volume_by_stock.get(sid, 0.0), existing_keys,
            )
            if sig:
                to_save.append(sig)

        await signal_repo.add_many(to_save)

        rapid = sum(1 for s in to_save if s.signal_type == SignalType.RAPID_DECLINE.value)
        trend = sum(1 for s in to_save if s.signal_type == SignalType.TREND_REVERSAL.value)
        squeeze = sum(1 for s in to_save if s.signal_type == SignalType.SHORT_SQUEEZE.value)
        elapsed = int((time.monotonic() - start) * 1000)
        logger.info(
            "시그널 탐지 완료 rapid=%d trend=%d squeeze=%d elapsed=%dms",
            rapid, trend, squeeze, elapsed,
        )
        return DetectionResult(
            rapid_decline=rapid, trend_reversal=trend, short_squeeze=squeeze, elapsed_ms=elapsed
        )

    # ------------------------------------------------------------------
    # 개별 탐지 로직
    # ------------------------------------------------------------------

    @staticmethod
    def _rapid_decline(
        stock_id: int, trading_date: date, lb: Any, existing: set[tuple[int, str]]
    ) -> Signal | None:
        if lb is None or lb.change_rate is None:
            return None
        if Decimal(lb.change_rate) > RAPID_DECLINE_THRESHOLD:
            return None
        key = (stock_id, SignalType.RAPID_DECLINE.value)
        if key in existing:
            return None
        existing.add(key)

        abs_change = abs(float(lb.change_rate))
        base = min(60, int(abs_change * 2.5))
        consec = min(20, int(lb.consecutive_decrease_days or 0) * 5)
        score = min(100, base + consec + 10)
        detail = {
            "balanceChangeRate": str(lb.change_rate),
            "changeQuantity": int(lb.change_quantity or 0),
            "consecutiveDecreaseDays": int(lb.consecutive_decrease_days or 0),
        }
        return Signal(
            stock_id=stock_id, signal_date=trading_date,
            signal_type=SignalType.RAPID_DECLINE.value,
            score=score, grade=_grade(score), detail=detail,
        )

    @staticmethod
    def _trend_reversal(
        stock_id: int, trading_date: date, hist: pd.DataFrame | None,
        existing: set[tuple[int, str]],
    ) -> Signal | None:
        if hist is None or len(hist) < TREND_MA_LONG + 1:
            return None
        sorted_hist = hist.sort_values("date").reset_index(drop=True)
        qty = sorted_hist["qty"]

        ma_short = qty.rolling(TREND_MA_SHORT, min_periods=TREND_MA_SHORT).mean()
        ma_long = qty.rolling(TREND_MA_LONG, min_periods=TREND_MA_LONG).mean()
        # rolling 결과는 NaN 이 나올 수 있고 None 은 절대 나오지 않음 — pd.isna 로 통일 검증.
        if pd.isna(ma_short.iloc[-1]) or pd.isna(ma_long.iloc[-1]):
            return None
        if pd.isna(ma_short.iloc[-2]) or pd.isna(ma_long.iloc[-2]):
            return None
        short_today = float(ma_short.iloc[-1])
        long_today = float(ma_long.iloc[-1])
        short_yest = float(ma_short.iloc[-2])
        long_yest = float(ma_long.iloc[-2])

        cross = short_yest >= long_yest and short_today < long_today  # Java 동일 조건
        if not cross:
            return None

        divergence = abs(short_today - long_today) / long_today * 100 if long_today else 0
        divergence_score = min(40, int(divergence * 10))
        speed = (
            abs((short_today - short_yest) / short_yest * 100) if short_yest else 0
        )
        speed_score = min(30, int(speed * 15))
        score = min(100, divergence_score + speed_score + 30)
        if score < TREND_REVERSAL_MIN_SCORE:
            return None

        key = (stock_id, SignalType.TREND_REVERSAL.value)
        if key in existing:
            return None
        existing.add(key)
        detail = {
            "maShort": round(short_today, 2),
            "maLong": round(long_today, 2),
            "crossType": "GOLDEN_CROSS",  # Java 용어 보존
        }
        return Signal(
            stock_id=stock_id, signal_date=trading_date,
            signal_type=SignalType.TREND_REVERSAL.value,
            score=score, grade=_grade(score), detail=detail,
        )

    @staticmethod
    def _short_squeeze(
        stock_id: int, trading_date: date, lb: Any, sp: Any, ss: Any,
        avg_volume: float, existing: set[tuple[int, str]],
    ) -> Signal | None:
        if lb is None or sp is None:
            return None

        # balance score
        change_rate = _to_decimal(lb.change_rate) or Decimal(0)
        balance_score = min(30, int(abs(float(change_rate)) * 1.5))

        # volume score
        today_vol = int(sp.volume or 0)
        ratio = (today_vol / avg_volume) if avg_volume > 0 and today_vol > 0 else 0.0
        volume_score = max(0, min(25, int((ratio - 1) * 12.5))) if ratio > 0 else 0

        # price score
        price_change = _to_decimal(sp.change_rate) or Decimal(0)
        price_val = float(price_change)
        price_score = min(25, int(price_val * 5)) if price_val > 0 else 0

        # short ratio score
        short_ratio = _to_decimal(ss.short_ratio) if ss else None
        short_score = min(20, int(float(short_ratio) * 2)) if short_ratio else 0

        total = balance_score + volume_score + price_score + short_score
        if total < SHORT_SQUEEZE_MIN_SCORE:
            return None

        key = (stock_id, SignalType.SHORT_SQUEEZE.value)
        if key in existing:
            return None
        existing.add(key)

        detail = {
            "balanceChangeRate": str(change_rate),
            "balanceScore": balance_score,
            "volumeChangeRate": round(ratio, 4),
            "volumeScore": volume_score,
            "priceChangeRate": str(price_change),
            "priceScore": price_score,
            "shortRatioScore": short_score,
            "consecutiveDecreaseDays": int(lb.consecutive_decrease_days or 0),
        }
        return Signal(
            stock_id=stock_id, signal_date=trading_date,
            signal_type=SignalType.SHORT_SQUEEZE.value,
            score=total, grade=_grade(total), detail=detail,
        )
